### PR TITLE
Account

### LINKED ISSUE #1073: Starname account functionality
[:man_dancing:](https://www.youtube.com/watch?v=Xs72-qjujts)

**Register**
- [x] domain name must be unique
- [x] account name must be unique within a domain scope
- [x] only configuration owner can register a domain without a superuser
- [x] anyone can register a domain with superuser
- [x] registering a domain creates an account with an empty name
- [x] only a domain admin can register account under a domain that has superuser
- [x] anyone can register an account under a domain without a superuser
- [x] a new account cannot be registered under an expired domain
- [x] iov domain can be registered (iov domain will migrate to a domain without a superuser)

**Extension Configuration**
- [x] global extension configuration can be updated by the configuration owner

**Targets**
- ~Basic target functionality woks the same way as it does in x/username~
- [x] Valid target blockchain ID and blockchain address rules are configured via extension configuration.
- [x] an account owner can change the list of targets

**Certificates**
- [x] account owner can add a certificate to a account
- ~account owner can add a certificate to an account in that domain~
- ~account owner can add a certificate to an account with empty name~
- [x] certificate of an expired account or an account that belongs to an expired domain cannot be added
- [x] certificate of an expired account or an account that belongs to an expired domain can be removed
- [x] changing domain admin does remove all certificates from accounts that belong to that domain
- [x] changing account owner removes all certificates from that account
- [x] changing account targets does not remove certificates from that account
- [x] a certificate (identified by its content) must not appear more than once in the list
- [x] there is a message that an account owner can delete a single certificate identified by sha256 hash of the content (frontent is already using sha256)

**Delete**
- [x] deletion of the empty name account is not possible
- [x] an account owner can delete an account
- [x] a domain admin can delete any account in that domain that has a superuser
- [x] deletion of a domain deletes all accounts that domain contained
- [x] deleting all accounts (flushing) does not delete the empty name account
- [x] deletion of a non-existing domain fails
- [x] a domain can be deleted by the domain admin that has a superuser

**Renew**
- [x] anyone can renew a domain
- [x] renewing a domain does not shorten its expiration time
- [x] anyone can renew an account
- [x] a domain is expired when its expiration date is in the past
- [x] an account is expired when either that account or the domain that account belongs to expiration date is in the past

**Transfer and expiry**
- [x] a domain ownership (admin address) can be changed (transferred) by the domain admin
- [x] expired domain ownership (domain admin) cannot be changed
- [x] expired account target cannot be changed
- [x] expired account can be deleted
- [x] empty name account cannot be transferred
- [x] expired account cannot be transferred
- [x] only account owner can transfer an account that belongs to a domain with no superuser
- [x] account owner cannot transfer ownership of an account that belongs to a domain with a superuser

**Tracking**

- [ ] be able to track the exchange and renewal of the starname owners through brokers (third_party_something) - purpose would be to know which broker sold which starnames and to whom

### PR DESCRIPTION
**Pull request by husio**
cmd/bnsd/x/accounts extension added
    
    A new `x/accounts` extension for managing accounts and domains was
    created. This is an improvement over existing `x/username` package which
    is meant to be migrated to the new extension.
    
    resolve #1073

### CODE REVIEW

File: cmd/bnsd/x/account/codec.proto [hunk 1]
```diff
@@ -97,26 +122,38 @@ message UpdateConfigurationMsg {
...
-  // Domain part of the account that should be registered.
-  string domain = 3;
+  string domain = 2;
+  // Admin holds the address of the newly registered domain.
+  bytes admin = 3 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+  // Has Superuser is a feature switch flag.
+  bool has_superuser = 4;
   // Third party token is an arbitrary data that can be provided by a middleman
   // (i.e. retailer) that although does not participate directly in the
   // transaction, helped a user to register a domain. Storing a client token
```
Comment (orkunkl) [inline cmd/bnsd/x/account/codec.proto]:
```suggestion
  // transaction, helped a user to register a domain. Storing a third party token
```

Comment (orkunkl) [inline cmd/bnsd/x/account/codec.proto]:
```suggestion
// transaction, helped a user to register a domain. Storing a third party token
```

Comment (orkunkl) [inline cmd/bnsd/x/account/codec.proto]:
```suggestion
  // transaction, helped a user to register a domain. Storing a third party token
```

File: cmd/bnsd/x/account/codec.proto [hunk 2]
```diff
@@ -190,8 +231,29 @@ message DeleteAccountMsg {
 
 // FlushDomainMsg issues deletion of all names (except the one with an empty
 // name) that belongs to given domain.
-// Message must be signed by the domain owner.
+// Message must be signed by the domain admin.
 message FlushDomainMsg {
```
Comment (orkunkl) [inline cmd/bnsd/x/account/codec.proto]: 🚽

File: cmd/bnsd/x/account/codec.proto [hunk 3]
```diff
@@ -190,8 +231,29 @@ message DeleteAccountMsg {
...
+// RenewAccountMsg is issuing an account expiration time extension. Extension
+// period is defined by the domain that this account belongs to.
+message RenewAccountMsg {
+  weave.Metadata metadata = 1;
+  // Name of the domain that is transferred.
+  string domain = 2;
+  string name = 3;
+  // Third party token is an arbitrary data that can be provided by a middleman
+  // (i.e. retailer) that although does not participate directly in the
+  // transaction, helped a user to renew a domain. Storing a client token helps
```
Comment (orkunkl) [inline cmd/bnsd/x/account/codec.proto]:
```suggestion
  // transaction, helped a user to renew a domain. Storing a third party token helps
```

File: spec/gogo/cmd/bnsd/x/account/codec.proto [hunk 4]
```diff
@@ -1,33 +1,46 @@
...
+  // required when processing a message within this domain.
+  repeated MsgFee msg_fees = 6 [(gogoproto.nullable) = false];
+  // Account review defines the duration of the account renewal period for each
+  // account that belongs to this domain.
+  int64 account_renew = 7 [(gogoproto.casttype) = "github.com/iov-one/weave.UnixDuration"];
+}
+
+// MsgFee defines a fee for processing a message of a given type, determined by
+// that message path.
+message MsgFee {
```
Comment (orkunkl) [inline spec/gogo/cmd/bnsd/x/account/codec.proto]: `MsgFee` naming could be confusing since there are modules filled with msg fee functionality.

**Comment by iov-bot-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: spec/gogo/cmd/bnsd/x/account/codec.proto (hunk 4 above)
Comment (husio) [reply spec/gogo/cmd/bnsd/x/account/codec.proto]: done in https://github.com/iov-one/weave/pull/1071/commits/40fd491bc079aeabc58c6ba761d74f3a9bb9f9f6

**Pull request merged**